src/classes/AFFL_Affiliations_TDTM.cls

```diff
@@ -53,34 +48,40 @@ public class AFFL_Affiliations_TDTM extends TDTM_Runnable {
-            
-            // we used to prevent any further recursive call with contacts and affiliations,
-            // but it turns out in some scenarios, before the AfterInsert of a contact calls us,
-            // another trigger handler may update the contact, causing the AfterUpdate to come thru.
-            // if we set the flag to stop recursion, then we won't process the AfterInsert, where
-            // we create the initial affiliation. thus we no longer set a recursion prevention flag,
-            // and we only use doNotRunForContacts where we actually have a use case to prevent recursion.
-            // as for the Affiliations and Account recursion flags, we have no current known issues with
-            // their recursion prevention, so we left that as is.
-            //
-            // hasRunForContacts = true;
-            
-            return wrapper;
-        } else if (objResult.getsObjectType() == npe5__Affiliation__c.sObjectType && !hasRunForAffiliations) {
-            DmlWrapper wrapper = runAfflTriggerHandler(newlist, oldlist, triggerAction, objResult);
-            hasRunForAffiliations = true;
-            return wrapper;
-        } else if (objResult.getsObjectType() == Account.sObjectType && !hasRunForAccounts) {
-            DmlWrapper wrapper = runAccountTriggerHandler(newlist, oldlist, triggerAction, objResult);
-            hasRunForAccounts = true;
-            return wrapper;
+        if (objResult.getsObjectType() == Contact.sObjectType && !TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_doNotRunForContacts)) {
+            wrapper = runContactTriggerHandler(newlist, oldlist, triggerAction, objResult);
+
+        } else if (objResult.getsObjectType() == npe5__Affiliation__c.sObjectType && !TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_hasRunForAffiliations)) {
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.AFFL_hasRunForAffiliations, true);
+            wrapper = runAfflTriggerHandler(newlist, oldlist, triggerAction, objResult);
+
+        } else if (objResult.getsObjectType() == Account.sObjectType && !TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_hasRunForAccounts)) {
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.AFFL_hasRunForAccounts, true);
+            wrapper = runAccountTriggerHandler(newlist, oldlist, triggerAction, objResult);
+
         } else {
             return null;
         }
+
+        // Process the DML
+        TDTM_TriggerHandler.processDML(wrapper);
+
+        // Decrement the recursion counters based on related processing logic
+        if (objResult.getsObjectType() == npe5__Affiliation__c.sObjectType) {
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.AFFL_hasRunForAffiliations, false);
+            if (TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_hasRunForAccounts)) {
```

Comment:
i don't understand this.  shouldn't the unsetting match the setting before the processDML?  why is this one nested and checking for the flag set on accounts and then clearing the flag on contacts?